// muka/packages/plugs/date.js
var formatDate = ['YYYY-MM', 'YYYY-MM-DD', 'YYYY-MMM-DD'];
var date = {
    getDate: function (dateVal) {
        var time = dateVal ? new Date(dateVal) : new Date;
        var month = time.getMonth() + 1;
        var day = time.getDate();
        var hours = time.getHours();
        var minutes = time.getMinutes();
        var seconds = time.getSeconds();
        return {
            year: time.getFullYear().toString(),
            month: month >= 1 && month <= 9 ? '0' + month : month.toString(),
            day: day >= 1 && day <= 9 ? '0' + day : day.toString(),
            hours: hours >= 1 && hours <= 9 ? '0' + hours : hours.toString(),
            minutes: minutes >= 1 && minutes <= 9 ? '0' + minutes : minutes.toString(),
            seconds: seconds >= 1 && seconds <= 9 ? '0' + seconds : seconds.toString(),
            milliSeconds: time.getMilliseconds().toString(),
            date: time
        };
    },
    getFormatDate: function (limiter, dateVal) {
        if (limiter === void 0) { limiter = 'YYYY-MM-DD'; }
        var time = this.getDate(dateVal);
        var o = {
            'MM': time.month,
            'DD': time.day,
            'hh': time.hours,
            'mm': time.minutes,
            'ss': time.seconds,
            'qq': Math.floor((time.date.getMonth() + 3) / 3),
            'S': time.milliSeconds // 毫秒
        };
        var week = {
            '0': '日',
            '1': '一',
            '2': '二',
            '3': '三',
            '4': '四',
            '5': '五',
            '6': '六'
        };
        if (/(Y+)/.test(limiter)) {
            limiter = limiter.replace(RegExp.$1, time.year.substr(4 - RegExp.$1.length));
        }
        if (/(E+)/.test(limiter)) {
            limiter = limiter.replace(RegExp.$1, ((RegExp.$1.length > 1) ? (RegExp.$1.length > 2 ? '星期' : '周') : '')
                + week[time.date.getDay() + '']);
        }
        for (var i in o) {
            if (new RegExp("" + i).test(limiter)) {
                console.log(11111);
                limiter = limiter.replace(i, o[i]);
            }
        }
        return limiter;
    },
    getNowNumber: function (dateVal) {
        var time = this.getDate(dateVal);
        return Number(time.year + time.month + time.day + time.hours + time.minutes + time.seconds + time.milliSeconds);
    }
};
export default date;
